File: Python/189-Rotate-Array.py
class Solution(object):
    def rotate(self, nums, k):
        length = len(nums)
        k %= length
        last = nums[-k:]
        for i in reversed(range(length - k)):
            nums[i+k] = nums[i]
        for i in range(len(last)):
            nums[i] = last[i]

File: JavaScript/189-rotate-array.js
let rotate = (nums, k) => {
  k %= nums.length
  reverse(nums, 0, nums.length - 1)
  reverse(nums, 0, k - 1)
  reverse(nums, k, nums.length - 1)
}

let reverse = (nums, start, end) => {
  while (start < end) [nums[start++], nums[end--]] = [nums[end], nums[start]]
}
